runtime/src/multisig.rs

```diff
@@ -0,0 +1,415 @@
+            ensure!(<MultiSigSignsRequired<T>>::exists(&sender), "Multi sig does not exist");
+            <MultiSigSignsRequired<T>>::insert(&sender, &sigs_required);
+            Self::deposit_event(RawEvent::MultiSigSignaturesRequiredChanged(sender, sigs_required));
+            Ok(())
+        }
+    }
+}
+
+decl_event!(
+    pub enum Event<T>
+    where
+        AccountId = <T as system::Trait>::AccountId,
+    {
+        /// Event for multi sig creation. (MultiSig address, Creator address, Signers(pending approval), Sigs required)
+        MultiSigCreated(AccountId, AccountId, Vec<Signer>, u64),
+        /// Event for adding a proposal (MultiSig, proposalid)
+        ProposalAdded(AccountId, u64),
+        /// Emitted when a proposal is executed. (MultiSig, proposalid, result)
+        ProposalExecuted(AccountId, u64, bool),
+        /// Signer added (Authorization accepted) (MultiSig, signer_added)
+        MultiSigSignerAdded(AccountId, Signer),
+        /// Multi Sig Signer Authorized to be added (MultiSig, signer_authorized)
+        MultiSigSignerAuthorized(AccountId, Signer),
+        /// Multi Sig Signer removed (MultiSig, signer_removed)
+        MultiSigSignerRemoved(AccountId, Signer),
+        /// Change in signatures required by a multisig (MultiSig, new_sigs_required)
+        MultiSigSignaturesRequiredChanged(AccountId, u64),
+    }
+);
+
+impl<T: Trait> Module<T> {
+    /// Creates a new proposal
+    pub fn create_proposal(
+        multisig: T::AccountId,
+        proposal: Box<T::Proposal>,
+        sender_signer: Signer,
+    ) -> Result {
+        ensure!(
+            Self::ms_signers((multisig.clone(), sender_signer)),
+            "not an signer"
+        );
+        let proposal_id = Self::ms_tx_done(multisig.clone());
+        <Proposals<T>>::insert((multisig.clone(), proposal_id), proposal);
+        // Since proposal_ids are always only incremented by 1, they can not overflow.
+        let next_proposal_id: u64 = proposal_id + 1u64;
+        <MultiSigTxDone<T>>::insert(multisig.clone(), next_proposal_id);
+        Self::deposit_event(RawEvent::ProposalAdded(multisig.clone(), proposal_id));
+        Self::approve_for(multisig, proposal_id, sender_signer)
+    }
+
+    /// Approves a multisig transaction and executes the proposal if enough sigs have been received
+    fn approve_for(multisig: T::AccountId, proposal_id: u64, signer: Signer) -> Result {
+        let multisig_signer_proposal = (multisig.clone(), signer, proposal_id);
+        let multisig_proposal = (multisig.clone(), proposal_id);
+        ensure!(!Self::votes(&multisig_signer_proposal), "Already approved");
+        if let Some(proposal) = Self::proposals(&multisig_proposal) {
+            Self::charge_fee(multisig.clone(), proposal.get_dispatch_info().weight)?;
+            <Votes<T>>::insert(&multisig_signer_proposal, true);
+            // Since approvals are always only incremented by 1, they can not overflow.
```

Comment:
They can, even though it'll take a lot of time ;)
`overflowing_add` is a simple register addition with carry. Why not just use it like it's common in nonce generation instead of an op that would crash the program on integer overflow?

Reply:
Billions of years of time :)

I don't think the default behavior of Rust is to crash on overflow. For performance reasons, rust doesn't check for overflow in release mode (can be toggled using a compiler flag though). It just wraps and returns two-compliment. This is why, unless we check what flag is returned by `overflowing_add`, it has no effect. Checking the flag is obviously a performance penalty.

Performance is the sole reason I am not doing checked arithmetic here. Substrate framework will give up much before we come to a point where this overflows since transaction nonce, block number etc use non-checked math and time is represented in u64.

Checked math and unchecked math can be the difference between 100 and 101 batch claims addition ;).

Reply:
You are right. I forgot that `+` crashes on overflow only in debug mode! It's totally fine to have `+` in that case.